src/content/language/extension-types.md

```diff
@@ -471,13 +471,21 @@ void main() {
 }
 ```
 
-It's important to be aware of this quality when using extension types.
+When `i` gets the static type `NumberE` based on a cast or a pattern match, `i`
+still refers to the same object, and `v` refers to the same object as `i`, and
+the object itself remains unchanged. It is only the static type that changed
+(and hence the methods that we can call on this object). In particular, the
+change of type does not involve an invocation of a constructor. If you wish to
+execute a constructor (for example, to perform some kind of validation) then it
+is necessary to write an explicit constructor invocation (such as `NumberE(i)`).
```

Comment:
Thanks for adding this additional clarification. I've seen incorrect assumptions around this behavior a few times, so I'm glad we call out it specifically now.